crates/doc/src/inference.rs

```diff
@@ -444,6 +445,82 @@ impl ObjShape {
             None
         }
     }
+
+    fn widen<'n, N>(&mut self, fields: &'n N::Fields, loc: Location, is_first_time: bool) -> bool
+    where
+        N: AsNode,
+    {
+        use crate::{Field, Fields};
+        // If a particular location defines `additionalProperties` as a subschema
+        // don't add the field from `AsNode`, instead jump straight to squashing
+        match self.additional.as_mut() {
+            // We only want to squash if your `additionalProperties`
+            // is set to something other than `false`.
+            Some(addl) if !addl.type_.eq(&types::INVALID) => {
+                fields.iter().enumerate().fold(false, |accum, (idx, node)| {
+                    accum || addl.widen(node.value(), loc.push_item(idx))
+                })
+            }
+            _ => {
+                let mut hint = false;
+
+                let shapes: Vec<_> = itertools::merge_join_by(
+                    self.properties.iter_mut(),
+                    fields.iter(),
+                    |prop, field| prop.name.cmp(&field.property().to_string()),
+                )
+                .filter_map(|eob| match eob {
+                    // Both the shape and node have this field, recursion time
+                    EitherOrBoth::Both(lhs, rhs) => {
+                        hint |= lhs.shape.widen(rhs.value(), loc.push_prop(rhs.property()));
+                        None
+                    }
+                    // Shape has a field that the node is missing, so let's make sure it's not marked as required
+                    EitherOrBoth::Left(mut lhs) => {
+                        lhs.is_required = false;
+                        None
+                    }
+                    // The Node has a field that the shape doesn't, let's add it
+                    EitherOrBoth::Right(rhs) => {
+                        let mut prop = ObjProperty {
+                            name: rhs.property().to_owned(),
+                            // A field can only be required if every single document we've seen
+                            // has that field present. This means that ONLY fields that exist
+                            // on the very first object we encounter for a particular location should
+                            // get marked as required, as any subsequent "new" fields
+                            // by definition did not exist on previous objects (and so cannot be required)
+                            // otherwise they would already be in the Shape
+                            // (and we would be in the `EoB::Both` branch).
+                            is_required: is_first_time,
+                            // Leave shape blank here, we're going to recur and expand it right below
+                            // Note: Shape starts out totally unconstrained (types::ANY) by default,
```

Comment:
no action: We should probably remove / refactor the `Default` and `new` impls -- they don't make logical sense, and are used only for tests.